docs/contrib/keycloak.md

```diff
@@ -0,0 +1,148 @@
+
+3. **Configuring REMS to Authenticate with Keycloak**
+    1. Modify your custom REMS config file (called `./services/rems/simple-config.edn` in this example). Tell REMS to use OIDC by modifying the `:authentication` parameter, and setting the values for some OIDC parameters specific to your keycloak instance. Remember that docker-compose will mount this file into your REMS container as `/rems/config/config.edn`.
+    ```
+    :authentication :oidc
+
+    ; The location of Keycloak's OIDC metadata for the realm to which REMS is client
+    :oidc-metadata-url "http://keycloak:8080/auth/realms/rems-idp/.well-known/openid-configuration"
+
+    ; REMS' client ID, as defined in Keycloak
+    :oidc-client-id "rems-client"
+
+    ; REMS' client secret, as generated in keycloak in step #8 above
+    :oidc-client-secret "0ed4b9e0-5ff0-446a-ab7b-82dac465b9d5"
+    ```
+
+4. **Spinning up REMS**
+    1. Add any other configuration that your REMS instance needs to your `simple-config.edn` file. For example, you will want to populate the `database-url` and `port` parameters.
+    ```
+    {:port 3000
+    :database-url "postgresql://db:5432/rems?user=rems&password=remspassword"
+    :search-index-path "/tmp/rems-search-index"
+    :authentication :oidc
+    :public-url "http://localhost:3001/"
+
+    :oidc-metadata-url "http://rp-keycloak:8080/auth/realms/rems-idp/.well-known/openid-configuration"
+
+    :oidc-client-id "rems-client"
+    :oidc-client-secret "0ed4b9e0-5ff0-446a-ab7b-82dac465b9d5"
+    }
+    ```
+    2. Build rems, which (at the time of writing) requires:
+        1. [Installing Leiningen](https://leiningen.org/)
+        2. Building the Leiningen jar: `lein uberjar`
+        3. Building the Docker container: `docker-compose build rems`
+        4. Migrating the test data: `docker-compose run --rm -e CMD="migrate;test-data" rems`
+    3. Run `docker-compose up rems`
+    4. Navigate to REMS, at http://localhost:3001/ in this example.
+
+5. **Testing Authentication**
+    1. To test authentication with Keycloak, add a test User in Keycloak by navigating to `Users` and clicking `Add user`. Give them a `username` and clicking `Save`.
+    2. Navigate to the new User's `Credentials` tab and set a `password` for them. Set the `Temporary` toggle to `OFF`.
+    3. In REMS, at http://localhost:3001/ in this example, click the `Login` button to be redirected to Keycloak. 
+    4. Access the account using `varchar`/`varchar`. You should be authenticated and redirected back to REMS.
+
+6. **Exporting the Realm**
+    1. You can export your realm and save it source control, to reduce development and testing time. Simply navigate to `Export` on the bottom-left of the page, set `Export clients` to `ON`, and click `Export`.
+    2. Next time you want to add this realm to keycloak, navigate to `Add Realm` and select your `realm-export.json` file for import. The `Name` will auto-populate.
+
+## OIDC Hostname issues
```

Comment:
I think I solved these by adding Keycloak to `/etc/hosts` so that I can use `keycloak` domain name also outside Docker networking. I think this is a pretty standard solution. Also you can configure REMS `:public-url` to another  host and port.

Reply:
Adding keycloak to `/etc/hosts` is common practice, but it somewhat defeats the purpose of containerizing Keycloak. It means that the developer needs to add documentation about adding the keycloak domain to `/etc/hosts` to their build instructions, and can be problematic when the user installing the app has strict Research IT.

This solution also breaks once you have multiple keycloaks, served at various ports, sometimes with varying hostnames.

I haven't tried solving this using REMS `:public-url`, thanks for the tip! I will try it out and report back.

Reply:
`:public-url` only solves the part of REMS however, not Keycloak. And REMS uses the URL that Keycloak reports as it is.

Maybe we can add editing `/etc/hosts` as one option, for those that do have admin priviledges on their machine. And then provide the two alternatives for others.